### ISSUE TITLE
Converting booleans into floats does not consistently yield exact 0 or 1

### ISSUE THREAD

**Issue by jaaack-wang** [pytorch/pytorch#91294]
### 🐛 Describe the bug

**Updated**: Problem solved by using `.isclose()`

I wrote a function to compute the full-sequence accuracy of predicted sequences against the target sequences. Part of the function looks like the following:

```python
import torch

def metrics(Y, Ypred):
    pw_cmp = (Y == Ypred).float()
    # batch-wise pair-wise overlap rate
    batch_overlap_rate = pw_cmp.mean(dim=0)
    
    # overlap_rate and absolute accuracy
    overlap_rate = batch_overlap_rate.mean().item()
    abs_correct = (batch_overlap_rate == 1.0)
    abs_accu = abs_correct.float().mean().item()
```

One thing that is puzzling me is that sometimes `abs_correct` will contain all `False` even though most of the values inside `batch_overlap_rate` is 1. I do not know why I got the following. Is it an error?

```python
print(batch_overlap_rate[0])
print(batch_overlap_rate2[0])
print(batch_overlap_rate[0] == torch.tensor(1.))
print(batch_overlap_rate2[0] == torch.tensor(1.))

# tensor(1.0000, device='cuda:0')
# tensor(1., device='cuda:0') 
# tensor(False, device='cuda:0')
# tensor(True, device='cuda:0')
```

If I print `batch_overlap_rate` and `batch_overlap_rate2` out, essentially both are a torch.Tensor of `torch.Size([1000])` and consists of lots of 1.0s, but comparing the former with 1.0 yields only `False`. I checked the data types of both variables. Nothing different. The function is fixed, and this issue only occurs **occasionally** but in a very particular way. Only sequences of certain lengths lead to this problem. There are two other metrics inside this function (e.g., overlap rate), their values are always right. 

**Updated**: The problem seems to be linked to floating precision. However, the output of `(Y == Ypred).float()` should always be either 0 or 1, right?

```python
print(batch_overlap_rate[0] - 1)
print(batch_overlap_rate2[0] - 1)

# tensor(-5.9605e-08, device='cuda:0')
# tensor(0., device='cuda:0')
```

### Versions

```
2022-12-22 05:45:51 (9.46 MB/s) - ‘collect_env.py’ saved [17278/17278]

Collecting environment information...
PyTorch version: 1.13.0+cu116
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.22.6
Libc version: glibc-2.27

Python version: 3.8.16 (default, Dec  7 2022, 01:12:13)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.10.133+-x86_64-with-glibc2.27
Is CUDA available: True
CUDA runtime version: 11.2.152
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 460.32.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.1.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.21.6
[pip3] torch==1.13.0+cu116
[pip3] torchaudio==0.13.0+cu116
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.14.0
[pip3] torchvision==0.14.0+cu116https://docs.github.com/github/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax
[conda] Could not collect
```

**Comment by drisspg** [comment]
I can not reproduce the issue you are providing. If you can produce a standalone script that reproduces the issue that would be very helpful. That being said the problem appears to be due to fp precision issues and any static casts that might be made when converting to and from tensors of different datatypes. I am going to close this issue for now.
